### ISSUE TITLE
poor rw performance when a database contains compressed tables 

### ISSUE THREAD

**Issue by vbp1** [orioledb/orioledb#25]
testing env: 2xCPU Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz (72 cores), 256GB RAM, 4x1.6TB NVME INTEL SSDPEDMD016T4 RAID0 for pgdata, Altlinux c9f1
orioledb specific settings:
```
orioledb.main_buffers = 16GB
orioledb.free_tree_buffers = 1GB
orioledb.catalog_buffers = 256MB
orioledb.undo_buffers = 1GB
orioledb.recovery_pool_size = 4
orioledb.recovery_queue_size = 64MB
orioledb.checkpoint_completion_ratio = 0.9
orioledb.bgwriter_num_workers = 8
orioledb.max_io_concurrency = 64
```
pgbench (Scale factor = 75К)  orioledb alpha5 + uncompressed tables vs orioledb branch fix_free_extents_vacated_bytes + compressed tables test results looks like that:
![pgbench_75k_rw](https://user-images.githubusercontent.com/60384539/162252287-1e51f530-aad4-4500-bf7c-20228539e101.png)

pgbench exec string was /opt/pgpro/ent-13/bin/pgbench --client=n --jobs=n --protocol=prepared --progress=10 --time=900 --username=pgbench --host=127.0.0.1 --port=5432 pgbench

I've noticed a lot of LWLock | BufferContent while taking the orioledb branch fix_free_extents_vacated_bytes + compressed tables test.

perf top:
Samples: 14K of event 'cycles', 99 Hz, Event count (approx.): 410994119338 lost: 0/0 drop: 10042/10042
Overhead  Shared Object         Symbol
  47.96%  orioledb.so           [.] get_extent
  12.67%  orioledb.so           [.] find_page
  10.23%  orioledb.so           [.] free_tree_len_off_cmp
   6.58%  orioledb.so           [.] btree_page_search
   2.62%  orioledb.so           [.] o_btree_read_page

**Comment by akorotkov** [comment]
Hi, @vbp1!
Thank you for reporting.
Could you please re-check this on 9aed798dbc?

**Comment by vbp1** [comment]
> Hi, @vbp1! Thank you for reporting. Could you please re-check this on [9aed798](https://github.com/orioledb/orioledb/commit/9aed798dbccb315b69e097a34659c1053fdd26b2)?

Seems like not good:
2022-05-04 09:10:23 MSK [3355058]: [233-1] us=,db=,app=,cli= LOG:  checkpoint starting: time
2022-05-04 09:10:23 MSK [3355058]: [234-1] us=,db=,app=,cli= LOG:  orioledb checkpoint 59 started
*** stack smashing detected ***: <unknown> terminated
2022-05-04 09:14:20 MSK [3354731]: [9-1] us=,db=,app=,cli= LOG:  server process (PID 2028408) was terminated by signal 11: Segmentation fault
2022-05-04 09:14:20 MSK [3354731]: [10-1] us=,db=,app=,cli= DETAIL:  Failed process was running: UPDATE pgbench_accounts SET abalance = abalance + $1 WHERE aid = $2;
2022-05-04 09:14:20 MSK [3354731]: [11-1] us=,db=,app=,cli= LOG:  terminating any other active server processes
2022-05-04 09:14:23 MSK [2066021]: [1-1] us=postgres,db=postgres,app=[unknown],cli=[local] FATAL:  the database system is in recovery mode
2022-05-04 09:14:25 MSK [3354731]: [12-1] us=,db=,app=,cli= LOG:  issuing SIGKILL to recalcitrant children
2022-05-04 09:14:26 MSK [3354731]: [13-1] us=,db=,app=,cli= LOG:  all server processes terminated; reinitializing

**Comment by akorotkov** [comment]
I see there should be core dump.

```
May 04 09:15:11 odb1.test.loc systemd-coredump[2065497]: Process 2028505 (postgres) of user 46 dumped core.
                                                         
                                                         Stack trace of thread 2028505:
                                                         #0  0x00007f3a0813ae83 n/a (n/a + 0x0)
```

Did you build postgres with debug symbols?  Any chance to get the stack trace?

**Comment by vbp1** [comment]
> I see there should be core dump.
> 
> ```
> May 04 09:15:11 odb1.test.loc systemd-coredump[2065497]: Process 2028505 (postgres) of user 46 dumped core.
>                                                          
>                                                          Stack trace of thread 2028505:
>                                                          #0  0x00007f3a0813ae83 n/a (n/a + 0x0)
> ```
> 
> Did you build postgres with debug symbols? Any chance to get the stack trace?

No. No in the moment
But I will try to reproduce the error and give you a trace.

**Comment by vbp1** [comment]
> I see there should be core dump.
> 
> ```
> May 04 09:15:11 odb1.test.loc systemd-coredump[2065497]: Process 2028505 (postgres) of user 46 dumped core.
>                                                          
>                                                          Stack trace of thread 2028505:
>                                                          #0  0x00007f3a0813ae83 n/a (n/a + 0x0)
> ```
> 
> Did you build postgres with debug symbols? Any chance to get the stack trace?

BTW how to build orioledb extension with debug symbols?

**Comment by akorotkov** [comment]
If PostgreSQL is build with `--enable-debug`, extension also gets debug symbols during build.

**Comment by akorotkov** [comment]
I would say it's much better already :)
We will keep improving this.  I will write about next commit to recheck.

**Comment by vbp1** [comment]
> I would say it's much better already :) We will keep improving this. I will write about next commit to recheck.

Of course! The perf gain is significant. But we are quite far from ideal yet;)
